content/learn/book/migration-guides/0.6-0.7/_index.md

```diff
@@ -0,0 +1,211 @@
++++
+title = "0.6 to 0.7"
+weight = 1
+sort_by = "weight"
+template = "book-section.html"
+page_template = "book-section.html"
+insert_anchor_links = "right"
+[extra]
+long_title = "Migration Guide: 0.6 to 0.7"
++++
+
+<!-- Github filter used to find the relevant PRs "is:pr label:C-Breaking-Change closed:>2022-02-01 [Merged by Bors]" -->
+
+### ParamSet for conflicting SystemParam
+
+<https://github.com/bevyengine/bevy/pull/2765>
+
+```rs
+// 0.6
+fn system(
+    mut transforms: QuerySet<(
+        QueryState<&mut Transform, With<Marker>>,
+        QueryState<&Transform>,
+    )>,
+) {
+    for transform in transforms.q1().iter() {
+        // ...
+    }
+}
+// 0.7
+fn system(
+    mut transforms: ParamSet<(
+        Query<&mut Transform, With<Marker>>,
+        Query<&Transform>
+    )>,
+) {
+    for transform in transforms.p1().iter() {
+        // ...
+    }
+}
+```
+
+### AliasedMutability
+
+<https://github.com/bevyengine/bevy/pull/4298>
+
+The `QueryEntityError` enum now has a `AliasedMutability` variant, and returns the offending entity.
+
+### Remove margins.rs
+
+<https://github.com/bevyengine/bevy/pull/4284>
+
+The `Margins` type was removed. To migrate, replace of `Margins` with `Rect`.
+
+### Remove face_toward.rs
+
+<https://github.com/bevyengine/bevy/pull/4277>
+
+The `FaceToward` trait was removed. To migrate, replace every occurrence of `Mat4::face_toward` to `Mat4::look_at_rh`.
+
+### `World::entities_mut` is now unsafe
+
+<https://github.com/bevyengine/bevy/pull/4093>
+
+```rs
+// 0.6
+world.entities_mut()
+
+// 0.7
+unsafe { world.entities_mut() }
+```
+
+### Mesh vertex buffer layouts
+
+<https://github.com/bevyengine/bevy/pull/3959>
+
+TODO
```

Comment:
~~~suggestion

Custom vertex attributes are now referenced by a `MeshVertexAttribute` rather than a simple string and `set_attribute` has been renamed to `insert_attribute` better reflect its behavior.

```rs
// 0.6
mesh.set_attribute("Vertex_Custom", VertexAttributeValues::Sint32x4(vec![]));

// 0.7
// Generate your own random identifier here.
// https://play.rust-lang.org/?gist=cc7e824724ba023e9bff25db35ef1f5e
pub const ATTRIBUTE_CUSTOM: MeshVertexAttribute =
    MeshVertexAttribute::new("Custom", 17351772347970238659, VertexFormat::Sint32x4);
mesh.insert_attribute(
    ATTRIBUTE_CUSTOM,
    VertexAttributeValues::Sint32x4(vec![]),
);
```

Vertex buffers no longer need to be manually laid out with offset and stride values in a `RenderPipelineDescriptor`.

```rs
// 0.6
let vertex_buffer_layout = VertexBufferLayout {
    array_stride: 20,
    step_mode: VertexStepMode::Vertex,
    attributes: vec![
        VertexAttribute {
            format: VertexFormat::Float32x3,
            offset: 0,
            shader_location: 0,
        },
        VertexAttribute {
            format: VertexFormat::Float32x2,
            offset: 12,
            shader_location: 1,
        },
    ],
};

// 0.7
let mut formats = vec![
    VertexFormat::Float32x3,
    VertexFormat::Float32x2,
];
let vertex_layout = VertexBufferLayout::from_vertex_formats(VertexStepMode::Vertex, formats);

~~~

This one might need some additional eyes / touchups.

Using `VertexBufferLayout::from_vertex_formats` **seemed** necessary when migrating code for this change the two times I have done it, but I'm not 100% it actually is.

Reply:
I think I'll put the vertex attribute stuff in a separate category, but other than that it looks ok to me